server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1200,17 +1216,26 @@ private static Set<String> innerResolve(Context context, List<String> expression
                     continue;
                 }
                 wildcardSeen = true;
-
                 final Map<String, IndexAbstraction> matches = matches(context, expression);
                 if (context.getOptions().allowNoIndices() == false && matches.isEmpty()) {
                     throw indexNotFoundException(expression);
                 }
-                Set<String> expand = expand(context, matches, expression);
-                if (add) {
-                    result.addAll(expand);
-                } else {
-                    result.removeAll(expand);
-                }
+                Collection<String> finalResult = result;
+                expandMatches(context, matches.values(), expression.startsWith("."), expanded -> {
+                    if (add) {
+                        finalResult.add(expanded);
+                    } else {
+                        finalResult.remove(expanded);
+                    }
+                });
+            }
+            if (result == null) {
+                result = expressions;
```

Comment:
Is there a potential issue with us returning the original list here instead of a null? Is it safer to just return an empty Set, or do we need the original expressions? I think there's a code path where a non-empty expressions list can still lead to a null result.

Reply:
I think this is safe, and returning an empty Set would not work.

This PR is **very** hard to follow as a whole, but I've tried to make every commit self-contained, and I think it would be easier to review the changes incrementally commit by commit, eg https://github.com/elastic/elasticsearch/pull/89311/commits/9b90714bd4c63ec2d1961c499f680c81a2dd1d79 for this change.

Reply:
To be clear, the [original code](https://github.com/elastic/elasticsearch/blob/1403ab318b14ad39030c6824237e1477a9392dd8/server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L1167) returned `null` when all the passed-in expression items existed, in which case the caller would detect the `null` and itself return the passed-in expression. It is safe to move the "if null return expression argument" condition inside this method because the expression argument cannot cannot be empty, see also: https://github.com/elastic/elasticsearch/commit/9b90714bd4c63ec2d1961c499f680c81a2dd1d79 .